[yaml] Add Beam YAML Examples and Getting started docs

Polber opened the pull request:
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

Polber commented:
R: @robertwb

robertwb commented on sdks/python/apache_beam/yaml/docs/README.md:
```diff
@@ -0,0 +1,295 @@
+# Beam YAML Developer Documentation
```
I think this level of documentation (for developers) is best kept as a living doc (which you already have) rather than checked into the repo.

robertwb commented on sdks/python/apache_beam/yaml/examples/test/map_to_fields_callable_test.py:
```diff
@@ -0,0 +1,36 @@
+# coding=utf-8
```
Having a separate test file for (nearly) every example seems overkill. Perhaps we could consolidate?

robertwb commented on sdks/python/apache_beam/yaml/examples/test/regex_matches_yaml_test.py:
```diff
@@ -0,0 +1,41 @@
+
+YAML_FILE = '../regex_matches.yaml'
+
+EXPECTED = [
```
Actually, it looks like these are all *very* formulaic. 

What if we put a comment 

`# Expected: ...`

right in the yaml file itself? This would be useful to the reader of the yaml file, and would obviate the need for all these separate test files (we would just do a glob for *.yaml in the directories). 

We could alternatively add an "expected" parameter to LogForTesting (which would do its verification via AssertThat, no need to even scrape stdout) if we didn't want to parse it out ourselves.

Polber replied:
I like the idea of putting the `# Expected` tag in the yaml file so that a user can see what the expected output is easily.

robertwb commented on sdks/python/apache_beam/yaml/examples/README.md:
```diff
@@ -0,0 +1,48 @@
+# Examples Catalog
+
+<!-- TOC -->
+* [Examples Catalog](#examples-catalog)
```
Long term (post 2.54) I we'll want these to live in the beam website, not here. But we can move them then.

robertwb commented on sdks/python/apache_beam/yaml/examples/transforms/aggregation/combine_sum.yaml:
```diff
@@ -0,0 +1,58 @@
+        language: python
+        group_by: fruit
+        combine:
+          total_quantity: 
```
Let's make this example combine more than just one field.

Polber replied:
I added another aggregated field, is that what you meant? Or group on multiple fields? Might be nice to have examples of both cases.

robertwb replied:
Yep, this is what I was thinking.

robertwb commented on sdks/python/apache_beam/yaml/examples/transforms/elementwise/explode.yaml:
```diff
@@ -0,0 +1,31 @@
+      name: Gardening plants
+      config:
+        elements:
+          - produce: ['🍓Strawberry', '🥕Carrot', '🍆Eggplant', '🍅Tomato', '🥔Potato']
```
Not sure how clear this example is. Maybe add a non-exploding field?

Polber replied:
Done.

robertwb commented on sdks/python/apache_beam/yaml/examples/wordcount_minimal.yaml:
```diff
@@ -0,0 +1,52 @@
+  transforms:
+    - type: ReadFromText
+      config:
+        path: gs://dataflow-samples/shakespeare/kinglear.txt
```
Does one need credentials to access this public file? (If so, add a comment.)

Polber replied:
Added a comment.

robertwb commented on sdks/python/apache_beam/yaml/examples/wordcount_minimal.yaml:
```diff
@@ -0,0 +1,74 @@
+              import re
+              def my_mapping(row):
+                return re.findall(r"[A-Za-z\']+", row.line.lower())
+          count: "1"
```
I wonder if it'd be easier to follow if we exploded the words, then added the "count" field.

Polber replied:
That is much cleaner. Fixed.

robertwb commented on sdks/python/apache_beam/yaml/examples/examples_test.py:
```diff
@@ -0,0 +1,118 @@
+
+  @classmethod
+  def parse_test_methods(cls, path):
+    files = glob.glob(os.path.join(path, r'*.yaml'))
```
Nit: maybe accept a glob string directly rather than assuming `path/*.yaml` to be more flexible?

Polber replied:
Fixed.

robertwb reviewed (commented):
This is looking pretty good. 

Let's remove the docs changes as we're moving those to the site and get these examples in.

robertwb commented:
Any update on this?

Polber commented:
@robertwb All comments addressed and rebased on master

Polber commented on sdks/python/apache_beam/yaml/examples/README.md:
```diff
@@ -0,0 +1,48 @@
+### Element-wise
+These examples leverage the built-in mapping transforms including `MapToFields`,
+`Filter` and `Explode`. More information can be found about mapping transforms
+[here](../docs/yaml_mapping.md).
```
I pointed to UDF section since that is where MapToFields lives

Polber commented on sdks/python/apache_beam/yaml/examples/wordcount_minimal.yaml:
```diff
@@ -0,0 +1,75 @@
+# see https://cloud.google.com/docs/authentication/external/set-up-adc for more
+# information
+#
+# This pipeline reads in a text file, maps all words to a value of "1", sums
```
@robertwb I refactored the example a bit to make it follow the logic more semantically. It also outputs `Row(word=..., count=...)` instead of `Row(output="word: count")`

Let me know what you think

robertwb reviewed (commented):
Looks good once we fix the failing tests and missing license headers.

robertwb commented:
Looks like you need to add an apache license preamble to sdks/python/apache_beam/yaml/examples/README.md to make RAT happy.

Polber commented:
> Looks like you need to add an apache license preamble to sdks/python/apache_beam/yaml/examples/README.md to make RAT happy.

Are the other tests known issues? I don't see how my PR is affecting huggingface and rowcoder tests

kennknowles commented:
Perhaps rebasing will get to green if there were issues on `master`?

robertwb commented:
I kicked off some re-runs, but I'm not sure if it always rebases on head. +1 to merging with master and re-pushing.

kennknowles commented:
I think it does not. The various merge and HEAD commits associated with a PR are, I think, only updated when the head commit of the PR changes. So reruns will be against the same commit. (I could have this wrong, of course)

Polber commented:
@robertwb 

Finally got the tests green.

The `RowCoderTest` imports `beam`, which imports the `examples_test.py` file I added, which *used* to import `main.py`. In `main.py`, there is the line `LogicalType.register_logical_type(MillisInstant)` - This overrided the `MicrosInstant` that `RowCoderTest` requires to pass. 

I adjusted the test to run `yaml_transform.expand_pipeline` instead, which fixed that issue. I also realized that the examples tests were not actually running on presubmits, so I added several fixes to get them running and green.

robertwb commented:
Thanks. Nice to finally get this in.
